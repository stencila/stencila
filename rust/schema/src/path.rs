use std::{
    collections::VecDeque,
    fmt::{self, Debug},
};

use derive_more::{Deref, DerefMut, IntoIterator};

use common::{
    eyre::{bail, OptionExt, Report, Result},
    serde::{Deserialize, Serialize},
    serde_json::{self},
};
use node_type::NodeProperty;

/// A slot in a node path: either a property identifier or the index of a vector.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[serde(untagged, crate = "common::serde")]
pub enum NodeSlot {
    Property(NodeProperty),
    Index(usize),
}

impl From<NodeProperty> for NodeSlot {
    fn from(value: NodeProperty) -> Self {
        Self::Property(value)
    }
}

impl From<usize> for NodeSlot {
    fn from(value: usize) -> Self {
        Self::Index(value)
    }
}

impl TryFrom<serde_json::Value> for NodeSlot {
    type Error = Report;

    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
        use serde_json::Value::*;
        match value {
            String(name) => Ok(Self::from(NodeProperty::try_from(name.as_str())?)),
            Number(index) => index
                .as_u64()
                .ok_or_eyre("Expected u64")
                .map(|value| Self::from(value as usize)),
            _ => bail!("Unhandled JSON value for `PatchSlot`"),
        }
    }
}

/// A path to reach a node from the root: a vector of [`PatchSlot`]s
///
/// A [`VecDeque`], rather than a [`Vec`] so that when applying operations in
/// a call to `patch` the front of the path can be popped off.
#[derive(
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Deref,
    DerefMut,
    IntoIterator,
    Serialize,
    Deserialize,
)]
#[serde(crate = "common::serde")]
#[derive(Default)]
pub struct NodePath(VecDeque<NodeSlot>);

impl NodePath {
    pub fn new() -> Self {
        Self::default()
    }
}

impl From<NodeProperty> for NodePath {
    fn from(value: NodeProperty) -> Self {
        Self::from(NodeSlot::from(value))
    }
}

impl From<usize> for NodePath {
    fn from(value: usize) -> Self {
        Self::from(NodeSlot::from(value))
    }
}

impl From<NodeSlot> for NodePath {
    fn from(value: NodeSlot) -> Self {
        Self(VecDeque::from([value]))
    }
}

impl<const N: usize> From<[NodeSlot; N]> for NodePath {
    fn from(value: [NodeSlot; N]) -> Self {
        Self(value.into())
    }
}

impl<const N: usize> From<[NodeProperty; N]> for NodePath {
    fn from(value: [NodeProperty; N]) -> Self {
        Self(value.into_iter().map(NodeSlot::from).collect())
    }
}

impl TryFrom<serde_json::Value> for NodePath {
    type Error = Report;

    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
        use serde_json::Value::*;
        match value {
            Number(..) | String(..) => Ok(Self::from(NodeSlot::try_from(value)?)),
            Array(array) => Ok(Self(VecDeque::from(
                array
                    .into_iter()
                    .map(NodeSlot::try_from)
                    .collect::<Result<Vec<_>>>()?,
            ))),
            _ => bail!("Unhandled JSON value for `PatchPath`"),
        }
    }
}

impl Debug for NodePath {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (index, slot) in self.iter().enumerate() {
            if index != 0 {
                f.write_str(".")?;
            }
            slot.fmt(f)?;
        }

        Ok(())
    }
}
