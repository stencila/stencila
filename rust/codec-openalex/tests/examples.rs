use std::path::PathBuf;

use glob::glob;
use insta::assert_json_snapshot;

use stencila_codec::{
    Codec,
    eyre::{Context, OptionExt, Result, eyre},
};
use stencila_codec_openalex::OpenAlexCodec;

/// Decode CSL-JSON into articles
#[tokio::test]
async fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*.json";

    for path in glob(&pattern)?.flatten() {
        let id = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(".json").map(String::from))
            .ok_or_eyre("should have .json suffix")?;

        let (node, ..) = OpenAlexCodec
            .from_path(&path, None)
            .await
            .wrap_err_with(|| eyre!("Unable to deserialize {id}.json"))?;

        assert_json_snapshot!(id, node, {
            ".commit" => "redacted"
        });
    }

    Ok(())
}
