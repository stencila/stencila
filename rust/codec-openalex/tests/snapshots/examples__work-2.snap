---
source: rust/codec-openalex/tests/examples.rs
expression: node
---
{
  "type": "Article",
  "id": "https://openalex.org/W4288041529",
  "doi": "10.1177/01445987221113122",
  "abstract": [
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "This research study examines the integration of wireless system and solar cells for detection and location of faults in power grids by implementing minimum number of Phasor Measurement Units (PMUs) after considering zero injection buses. A fault in the grid networks indicates a failure in the entire network that interfaces the normal flow of current which can be measured using wireless systems. The fault location in power grid is problematic because of high existence in several lateral tap-offs. In addition, the identification of exact location with presence of faults in the power system and severity of the faults are challenging tasks due to bulk and complex nature of the electrical power system state estimation. The major requirement of PMU placement approach is to find the minimum number of PMUs and their location with maximum wireless coverage under fault conditions with a prerequisite for the projected algorithm. The integrated algorithm is tested and validated for different fault conditions where the results are compared with different conventional methods using different optimization models such as heuristics, decision rules and mathematical programming where the outcomes proves to be much effective in terms of number of PMU installation for proper placements and detection, location in number of fault systems thus achieving 82% reliability using Ant Lion Optimizer (ALO)."
          }
        }
      ]
    }
  ],
  "authors": [
    {
      "type": "Person",
      "name": "Hariprasath Manoharan"
    },
    {
      "type": "Person",
      "orcid": "0000-0003-4297-3460",
      "affiliations": [
        {
          "type": "Organization",
          "name": "University of Sheffield",
          "ror": "05krs5044"
        }
      ],
      "name": "Yuvaraja Teekaraman"
    },
    {
      "type": "Person",
      "name": "Ramya Kuppusamy"
    },
    {
      "type": "Person",
      "orcid": "0000-0003-3700-2491",
      "affiliations": [
        {
          "type": "Organization",
          "name": "Jimma University",
          "ror": "05eer8g02"
        }
      ],
      "name": "Arun Radhakrishnan"
    }
  ],
  "datePublished": {
    "type": "Date",
    "value": "2022-07-26"
  },
  "title": [
    {
      "type": "Text",
      "value": {
        "string": "Application of solar cells and wireless system for detecting faults in phasor measurement units using non-linear optimization"
      }
    }
  ],
  "content": [],
  "isPartOf": {
    "type": "PublicationIssue",
    "isPartOf": {
      "type": "PublicationVolume",
      "isPartOf": {
        "type": "Periodical",
        "name": "Energy Exploration & Exploitation"
      },
      "volumeNumber": "41"
    },
    "issueNumber": "1"
  },
  "pageStart": "210",
  "pageEnd": "223"
}
